[git](https://github.com/bdemeshev/coursera_linal/blob/master/week_01/problems/quiz_questions.yml)

# Задача 03

## Вариант 1
Для преобразования $$A: \begin{pmatrix} x_{1} \\ x_{2} \end{pmatrix} \rightarrow \begin{pmatrix} x_{1}+3 x_{2} \\ x_{1}-4 x_{2} \end{pmatrix} $$ Найдите все вектора $v = \begin{pmatrix} v_{1} \\ v_{2} \end{pmatrix} $, для которых $v \perp Av$. В ответе укажите отношение $v_{2}/ v_{1}$ (ответ округлите до двух знаков после запятой). Если получено несколько соотношений, для которых выполняется $v \perp Av$, укажите большее из них.

## Решение

$$v_{1} ( v_{1}+3 v_{2} ) + v_{2} (v_{1}-4 v_{2}) = 0$$ $$v_{1}^2 + 4 v_{1} v_{2} -4 v_{2}^2 = 0$$ $$ v_{2}/ v_{1} = 1/2 \pm \sqrt{2}/2 $$

In [ ]:
#@title Ответ
# range:(1.15, 1.3)

import numpy as np
.5 + 2**.5 / 2
#

1.2071067811865475

# Подробно
Чтобы найти все векторы $ v = (v_1, v_2) $, для которых $ v \perp Av $, сначала запишем $ A(v) $ для произвольного вектора $ v $.

$$ A(v) = A \begin{pmatrix} v_1 \\ v_2 \end{pmatrix} = \begin{pmatrix} v_1 + 3v_2 \\ v_1 - 4v_2 \end{pmatrix} $$

Вектор $ v \perp Av $ означает, что их скалярное произведение равно нулю:

$$ v \cdot A(v) = 0 $$

То есть:

$$ v_1(v_1 + 3v_2) + v_2(v_1 - 4v_2) = 0 $$

Раскроем скобки и упростим выражение:

$$ v_1^2 + 3v_1v_2 + v_1v_2 - 4v_2^2 = 0 $$
$$ v_1^2 + 4v_1v_2 - 4v_2^2 = 0 $$

Разделим на $ v_1^2 $ (предполагая, что $ v_1 \neq 0 $):

$$ 1 + 4 \frac{v_2}{v_1} - 4 \left(\frac{v_2}{v_1}\right)^2 = 0 $$

Обозначим $ t = \frac{v_2}{v_1} $. Получим квадратное уравнение относительно $ t $:

$$ 4t^2 - 4t - 1 = 0 $$

Получаем два корня:

$$ t_1 = \frac{1 + \sqrt{2}}{2} $$
$$ t_2 = \frac{1 - \sqrt{2}}{2} $$

Из двух значений $ t $, большее из них:

$$ \boxed{1.21} $$

In [ ]:
# TODO

import numpy as np

# Определяем матрицу преобразования
A = np.array([[1, 3], [1, -4]])

# Задаем систему уравнений, которые представляют условие перпендикулярности v и Av
# Мы ищем v = [v1, v2], такое что v^T * A * v = 0
# Это дает уравнение v1^2 + 4*v1*v2 - 4*v2^2 = 0

# Пусть v1 = 1 (можем взять любое ненулевое значение для v1)
v1 = 1

# Подставляем v1 в уравнение и находим v2
coeffs = [4, 4, -1]  # коэффициенты квадратного уравнения 4v2^2 + 4v2 - 1 = 0
v2_solutions = np.roots(coeffs)

# Вычисляем отношение v2/v1 для каждого решения и выбираем большее по модулю
ratios = abs(v2_solutions / v1)
max_ratio = np.max(ratios)

print(f"Максимальное отношение v2/v1: {max_ratio:.2f}")


Максимальное отношение v2/v1: 1.21
